File: ObserverPatternTest/app/src/main/java/charlezz/observerpatterntest/MainActivity.java
package charlezz.observerpatterntest;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    Button btn;

    private ArrayList<OnCustomClickListener> mOnCustomClickListener = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        btn = (Button) findViewById(R.id.btn);
        btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                for (OnCustomClickListener listener : mOnCustomClickListener) {
                    listener.onClick();
                }
            }
        });

        Subscriber subscriber1 = new Subscriber("subscriber1");
        Subscriber subscriber2 = new Subscriber("subscriber2");

        addCustomClickListener(subscriber1);
        addCustomClickListener(subscriber2);

    }

    public void addCustomClickListener(OnCustomClickListener listener) {
        mOnCustomClickListener.add(listener);
    }

    public void removeCustomClickListener(OnCustomClickListener listener) {
        mOnCustomClickListener.remove(listener);
    }


}

File: KotlinFirst/app/src/main/java/charlezz/com/kotlinfirst/MainActivity.kt
package charlezz.com.kotlinfirst

import android.os.Bundle
import android.os.Looper
import android.support.v7.app.AppCompatActivity
import android.util.Log
import kotlinx.coroutines.experimental.async
import org.jetbrains.anko.*
import org.jetbrains.anko.coroutines.experimental.bg
import org.jetbrains.anko.sdk25.coroutines.onClick

class MainActivity : AppCompatActivity(), AnkoLogger {


    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
//
//        val view = verticalLayout {
//            val name = editText()
//            button("Say Hello"){
//                onClick{
//                    toast("Hello ${name.text}")
//                }
//            }
//        }
//        setContentView(view)



        MyActivityUI().setContentView(this)
        Log.e(TAG, MyClass.getInstance().hello)

        bg {
            if (mainLooper == Looper.myLooper()) {
                Log.e(TAG, "MainThread")
            } else {
                Log.e(TAG, "Not MainThread")
            }
        }

        async(kotlinx.coroutines.experimental.android.UI) {
            var count =0
            for (i in 1..10) {

                bg {
                    count ++


                    Log.e(TAG, "go sleep")
                    Thread.sleep(5000)
                    Log.e(TAG, "end sleep")
                }.await()

                if (mainLooper == Looper.myLooper()) {
                    Log.e(TAG, "MainThread"+count)
                } else {
                    Log.e(TAG, "Not MainThread1"+count)
                }



            }

            Log.e(TAG, "finished")
        }


    }

    private fun someMethod() {
    }
}

class MyActivityUI : AnkoComponent<MainActivity> {

    override fun createView(ui: AnkoContext<MainActivity>) = ui.apply {
        verticalLayout {
            val name = editText()
            button("Say Hello") {
                onClick { ctx.toast("Hello, ${name.text}!") }
            }
            val tv = textView() {
                text = "hello worl2d"
                textSize = 30.0f
            }

        }
    }.view
}